vantage-agent/lib/charms/vantage_agent/v0/agent_snapper.py

```diff
@@ -0,0 +1,193 @@
+        logger.debug(f"## Processing start event for {self._snap_name}.")
+        self.run_snap_service("start")
+        self._charm.unit.status = ActiveStatus()
+
+    def handle_stop(self, event: StopEvent):
+        logger.debug(f"## Processing stop event for {self._snap_name}.")
+        self.run_snap_service("stop")
+        self._charm.unit.status = MaintenanceStatus()
+
+    def handle_remove(self, event: RemoveEvent):
+        logger.debug(f"## Processing remove event for {self._snap_name}.")
+        self.remove_snap()
+
+    ## Operations
+    def is_snap_installed(self) -> bool:
+        cmd = [
+            self._snap_path,
+            "list",
+            self._snap_name,
+        ]
+
+        try:
+            subprocess.check_output(cmd)
+            return True
+        except subprocess.CalledProcessError:
+            return False
+
+    def is_snap_active(self) -> bool:
+        logger.debug(f"### Checking active status for {self._snap_name}.daemon")
+        cmd = [
+            self._snap_path,
+            "services",
+            f"{self._snap_name}.daemon",
+        ]
+
+        try:
+            status_output = subprocess.check_output(cmd).decode("utf-8")
+            status_line = status_output.split("\n")[1]
+            status = status_line.split()[2]
+            return status == "active"
+        except subprocess.CalledProcessError:
+            return False
+
+    def install_snap(self) -> None:
+        channel = self._charm.config["snap-channel"]
+        assert isinstance(channel, str)
+
+        confinement = self._charm.config["snap-confinement"]
+        assert isinstance(confinement, str)
+
+        if not self.is_snap_installed():
+            logger.debug(f"### Installing {self._snap_name}.")
+            self._sys_exec(
+                self._snap_path,
+                "install",
+                "--channel",
+                channel,
+                f"--{confinement}",
+                self._snap_name,
+            )
+        else:
+            logger.debug(f"### Refreshing {self._snap_name} (already installed).")
+            self._sys_exec(
+                self._snap_path,
+                "refresh",
+                "--channel",
+                channel,
+                f"--{confinement}",
+                self._snap_name,
+            )
+
+    def configure_snap(self) -> None:
+        logger.debug(f"### Configuring {self._snap_name}.")
+        raw_config = self._charm.config["snap-config"]
+        assert isinstance(raw_config, str)
+
+        snap_config = {}
+        for line in raw_config.splitlines():
+            if line.startswith("#"):
+                continue
+            key, value = line.split("=", 1)
+            snap_config[key.strip()] = value.strip()
+
+        for key, value in snap_config.items():
+            self._sys_exec(
+                self._snap_path,
+                "set",
+                self._snap_name,
+                f"{key}={value}",
+            )
```

Comment:
thinking about this logic, I think we should first run `snap unset ...` in all configurations and then run `snap set`. This would make sure that only the configurations defined in the config of the charm will be applied to the snap indeed.

Reply:
Resolved in commit e15fece.